File: tic_tac_toe.R
library(stringr)
suppressMessages(library(data.table))

select_player <- function(con) {
  cat("Please select x or o?  ")
  symbol <- readLines(con = con, n = 1)
  symbol <- as.character(symbol)

  while (!(symbol %in% c("X", "O", "x", "o"))) {
    cat("Please choose either 'x' or 'o' ")
    symbol <- readLines(con = con, n = 1)
    symbol <- as.character(symbol)
  }
  return(symbol)
}

display_board <- function(pos, check, comp_move = FALSE) {
  dim <- sqrt(length(pos))
  board <- data.table(matrix(data = pos, ncol = dim, nrow = dim))

  cat("\n")
  if (check == 2) {
    cat("Final board:")
  } else {
    (
      cat("Current board:")
    )
  }

  cat("\n")
  Sys.sleep(1)
  for (n in 1:dim(board)[1]) {
    if (n == dim(board)[1]) {
      cat(" ", board[[n, 1]], "|", board[[n, 2]], "|", board[[n, 3]], "\n")
    } else {
      cat(" ", board[[n, 1]], "|", board[[n, 2]], "|", board[[n, 3]], "\n")
      cat(" ---+---+---", "\n")
    }
  }
  cat("\n")
  if (check != 2 & comp_move == FALSE) {
    Sys.sleep(1.5)
    cat("Please select one of the open positions above ")
  }
}

select_position <- function(con) {
  input_pos <- readLines(con = con, n = 1)

  while (!(input_pos %in% c(1:9))) {
    cat("Invalid input. Please choose a number between 1 and 9:  ")
    input_pos <- readLines(con = con, n = 1)
  }

  input_pos <- as.integer(input_pos)

  return(input_pos)
}

invalid_check <- function(pos, input_pos) {
  open_pos <- which(!(pos == "x" | pos == "o"))
  if (length(open_pos) == length(pos)) {
    return(input_pos)
    break
  } else {
    while (!(input_pos %in% open_pos)) {
      cat("Oh no! That position is invalid, please select another position:  ")
      input_pos <- select_position(con)
    }
    return(input_pos)
  }
}

confirm_pos <- function(symbol, input_pos) {
  cat(sprintf("Place '%s' at position %s? [y/n] ", symbol, input_pos))
  answer <- readLines(con = con, n = 1)

  while (!(answer %in% c("y", "n"))) {
    cat("Invalid input. Please select y or n:  ")
    answer <- readLines(con = con, n = 1)
  }

  answer <- as.character(answer)
  return(answer)
}

update_position <- function(input_pos, pos, symbol) {
  input_pos <- invalid_check(pos, input_pos)
  answer <- confirm_pos(symbol, input_pos)
  if (answer == "y") {
    pos[input_pos] <- symbol
  } else {
    while (answer == "n") {
      cat("Please select another position  ")
      input_pos <- select_position(con)
      input_pos <- invalid_check(pos, input_pos)
      answer <- confirm_pos(symbol, input_pos)
    }
    pos[input_pos] <- symbol
  }
  return(pos)
}

computer_move <- function(pos, symbol, check) {
  open_pos <- which(!(pos == "x" | pos == "o"))
  if (length(open_pos) > 0 & check != 2) {
    cat("Computer moved ")
    cat("\n")
    move <- sample(open_pos, 1)

    if (symbol == "X") {
      comp_sym <- "O"
    } else if (symbol == "x") {
      comp_sym <- "o"
    } else if (symbol == "O") {
      comp_sym <- "X"
    } else if (symbol == "o") {
      comp_sym <- "x"
    }

    pos[move] <- comp_sym

    return(pos)
  } else {
    return(pos)
  }
}

check_winner <- function(pos) {
  open_pos <- which(!(pos == "x" | pos == "o"))
  win_combos <- list(
    pos[1:3], pos[4:6], pos[7:9],
    c(pos[c(1, 4, 7)]), c(pos[c(2, 5, 8)]), c(pos[c(3, 6, 9)]),
    c(pos[c(1, 5, 9)]), c(pos[c(3, 5, 7)])
  )
  for (win in win_combos) {
    check <- sum(duplicated(win))

    if (check == 2) {
      Sys.sleep(1.5)
      cat("\n")
      cat(sprintf("Player %s wins!", win[[1]]))
      break
    }
  }
  if (length(open_pos) == 0 & check != 2) {
    Sys.sleep(1.5)
    cat("\n")
    cat("It's a draw!")
    check <- 2
  }
  return(check)
}

if (interactive()) {
  con <- stdin()
} else {
  con <- "stdin"
}

cat("Welcome to Taylor's tic tac toe showdown!\n")
cat("\n")
Sys.sleep(1.5)
numRows <- 3
numCols <- 3
pos <- as.numeric(1:(numRows * numCols))

symbol <- select_player(con)
check <- check_winner(pos)
Sys.sleep(1.5)
display_board(pos, check)
while (check != 2) {
  input_pos <- select_position(con)
  pos <- update_position(input_pos, pos, symbol)
  if (check != 2) {
    check <- check_winner(pos)
    if (check == 2) {
      Sys.sleep(1.5)
      display_board(pos, check)
      break
    }
  }
  Sys.sleep(1.5)
  display_board(pos, check, comp_move = TRUE)
  Sys.sleep(1.5)
  pos <- computer_move(pos, symbol, check)
  if (check != 2) {
    check <- check_winner(pos)
    if (check == 2) {
      Sys.sleep(1.5)
      display_board(pos, check)
      break
    }
  }
  Sys.sleep(1.5)
  display_board(pos, check)
}

File: tic_tac_toe.py
import numpy as np
import sys
import time

def select_player():
    symbol = input("Please select X or O: ")
    symbol = symbol.upper()
    
    while symbol not in ["X", "O"]:
        symbol = input("Please choose either 'X' or 'O': ")
        symbol = symbol.upper()
    
    return symbol

def display_board(pos, check, comp_move=False):
    dim = int(np.sqrt(len(pos)))
    board = np.array(pos).reshape((dim, dim))
    
    print()
    if check == 2:
        print("Final board:")
    else:
        print("Current board:")
    
    print()
    sys.stdout.flush()
    for row in range(dim):
        if row == dim - 1:
            print(f" {board[row, 0]} | {board[row, 1]} | {board[row, 2]} ")
        else:
            print(f" {board[row, 0]} | {board[row, 1]} | {board[row, 2]} ")
            print(" ---+---+--- ")
    
    print()
    sys.stdout.flush()
    if check != 2 and not comp_move:
        print("Please select one of the open positions above")
    
    time.sleep(1.5)

def select_position():
    input_pos = input("Enter a number between 1 and 9: ")
    
    while input_pos not in map(str, range(1, 10)):
        input_pos = input("Invalid input. Please choose a number between 1 and 9: ")
    
    return int(input_pos)

def invalid_check(pos, input_pos):
    open_pos = [i + 1 for i, val in enumerate(pos) if val not in ["X", "O"]]
    if len(open_pos) == len(pos):
        return input_pos
    else:
        while input_pos not in open_pos:
            input_pos = select_position()
            input_pos = invalid_check(pos, input_pos)
        return input_pos

def confirm_pos(symbol, input_pos):
    answer = input(f"Place '{symbol}' at position {input_pos}? [y/n]: ")
    
    while answer not in ["y", "n"]:
        answer = input("Invalid input. Please select 'y' or 'n': ")
    
    return answer

def update_position(input_pos, pos, symbol):
    input_pos = invalid_check(pos, input_pos)
    answer = confirm_pos(symbol, input_pos)
    
    if answer == "y":
        pos[input_pos - 1] = symbol
    else:
        while answer == "n":
            print("Please select another position")
            input_pos = select_position()
            input_pos = invalid_check(pos, input_pos)
            answer = confirm_pos(symbol, input_pos)
        pos[input_pos - 1] = symbol
    
    return pos

def computer_move(pos, symbol, check):
    open_pos = [i + 1 for i, val in enumerate(pos) if val not in ["X", "O"]]
    if len(open_pos) > 0 and check != 2:
        print("Computer moved")
        move = np.random.choice(open_pos)
        
        if symbol == "X":
            comp_sym = "O"
        elif symbol == "O":
            comp_sym = "X"
        
        pos[move - 1] = comp_sym
    
    return pos

def check_winner(pos):
    open_pos = [i + 1 for i, val in enumerate(pos) if val not in ["X", "O"]]
    win_combos = [
        pos[:3], pos[3:6], pos[6:9],
        pos[::3], pos[1::3], pos[2::3],
        pos[::4], pos[2:8:2]
    ]
    
    for win in win_combos:
        check = sum(np.unique(win, return_counts=True)[1] > 1)
        if check == 2:
            print(f"\nPlayer {win[0]} wins!")
            sys.stdout.flush()
            return check
    
    if len(open_pos) == 0 and check != 2:
        print("\nIt's a draw!")
        sys.stdout.flush()
        check = 2
    
    return check

print("Welcome to Taylor's tic tac toe showdown!\n")
sys.stdout.flush()

pos = list(range(1, 10))
symbol = select_player()
check = check_winner(pos)
time.sleep(1.5)
display_board(pos, check)
while check != 2:
    input_pos = select_position()
    pos = update_position(input_pos, pos, symbol)
    
    if check != 2:
        check = check_winner(pos)
        if check == 2:
            time.sleep(1.5)
            display_board(pos, check)
            break

    time.sleep(1.5)
    display_board(pos, check, comp_move=True)
    time.sleep(1.5)
    pos = computer_move(pos, symbol, check)
    
    if check != 2:
        check = check_winner(pos)
        if check == 2:
            time.sleep(1.5)
            display_board(pos, check)
            break
        
    time.sleep(1.5)
    display_board(pos, check)
